### ISSUE TITLE
Finalise Source trait

### ISSUE THREAD

**Issue by alteous** [alteous/gltf#34]
The final design of the `Source` trait needs to be agreed upon prior to the planned 1.0 release.

The final design should handle at least:

- [x] User extensions (assumed to be handled due to the generality of `Source::source_external_data`)
- [x] Async I/O (achieved with futures)

Edit: https://github.com/alteous/gltf/issues/31 discusses whether a `Source` implementation should be mutable, which is an interesting consideration.

**Comment by bwasty** [comment]
Do you think Async I/O can be handled with the same trait?

**Comment by alteous** [comment]
I'm really not sure! Experimentation is required.

**Comment by fst3a** [comment]
What do you mean by async io and why is it needed?

**Comment by bwasty** [comment]
@fst3a It means non-blocking I/O and could be used to load multiple buffers/images in parallel.
See also https://aturon.github.io/blog/2016/08/11/futures/ and the discussion in #13 (it's a bit spread out, just search for `async` on the page).

**Comment by alteous** [comment]
I've been considering leveraging the [`AsyncRead`](https://docs.rs/tokio-io/0.1.2/tokio_io/trait.AsyncRead.html) trait of tokio-io and calling [`read_to_end`](https://docs.rs/tokio-io/0.1.2/tokio_io/io/fn.read_to_end.html) on whatever is returned from `read_gltf` / `read_external_data`. The trouble is the implementors of this trait are currently limited to trivial cases like `Box<T>` where `T: AsyncRead` - not helpful.

**Comment by alteous** [comment]
Using the [`AsyncRead`](https://docs.rs/handy_io/0.1.2/handy_io/io/trait.AsyncRead.html) trait from `handy_io`, I've conjured a default async version of the `Source` trait.

```rust
use futures::future::{err, lazy, ok, BoxFuture, Future};
use handy_io::io::AsyncRead;

pub trait Source: fmt::Debug {
    /// Error type.
    type Err: error::Error + fmt::Debug + From<io::Error>;
    
    /// Read the contents of a .gltf or .glb file.
    fn read_gltf(&self) -> Result<Box<Read + Send>, Self::Err>;

    /// Read the contents of external data.
    fn read_external_data(&self, uri: &str) -> Result<Box<Read + Send>, Self::Err>;

    /// Read the contents of a .gltf or .glb file asynchronously.
    fn async_read_gltf(&self) -> Result<BoxFuture<Vec<u8>, io::Error>, Self::Err> {
        let reader = self.read_gltf()?;
        lazy(move || ok(reader))
            .and_then(|reader| reader.async_read(vec![]))
            .then(|result| match result {
                Ok((_, data, _)) => ok(data),
                Err((_, _, error)) => err(error),
            })
            .boxed()
    }

    /// Read the contents of external data asynchronously.
    fn async_read_external_data(
        &self,
        uri: &str,
    ) -> Result<BoxFuture<Vec<u8>, io::Error>, Self::Err> {
        let reader = self.read_gltf()?;
        lazy(move || ok(reader))
            .and_then(|reader| reader.async_read(vec![]))
            .then(|result| match result {
                Ok((_, data, _)) => ok(data),
                Err((_, _, error)) => err(error),
            })
            .boxed()
    }
}
```

**Comment by alteous** [comment]
The above is probably not a good idea since the reader does not know when to finish reading. There doesn't seem to be any cross-platform abstractions for file based async I/O on crates.io. Perhaps instead we should provide an `AsyncSource` trait instead which is used only in `StaticImporter::async_import`:

```rust
use futures::BoxFuture;

pub trait AsyncSource: Debug {
    /// Associated error type.
    type Error: Debug + Error; 

    /// Read the contents of a .gltf or .glb file asynchronously.
    fn async_read_gltf(&self) -> BoxFuture<Vec<u8>, Self::Error>;

    /// Read the contents of external data asynchronously.
    fn async_read_external_data(&self, uri: &str) -> BoxFuture<Vec<u8>, Self::Error>;
}

impl StaticImporter {
    pub fn async_import_from_source<S>(&self, source: S) -> BoxFuture<Gltf<'static>, Error<S>>
        where S: AsyncSource
    {
        unimplemented!()
    }

    // other members omitted
}

**Comment by alteous** [comment]
On second thoughts, perhaps the `Source` trait and importers should only return futures? The advantage is flexibility of user implementation, since the user can unconditionally choose what calls are blocking, and flexibility of consummation, since importers can choose to wait on the returned futures.

The design would then like something like this:

```rust
use futures::{future, BoxFuture};

trait Source {
    type Error;
    fn gltf(&self) -> BoxFuture<Vec<u8>, Self::Error>;
    fn buffer(&self, uri: &str) -> BoxFuture<Vec<u8>, Self::Error>;
    fn image(&self, uri: &str, mime_type: &str) -> BoxFuture<Vec<u8>, Self::Error>;
}

impl StaticImporter {
    fn import<S: Source>(&self, source: S) -> BoxFuture<Gltf<'static>, Error<S>> {
        future::ok(unimplemented!())
    }

    fn import_sync<S: Source>(&self, source: S) -> Result<Gltf<'static>, Error<S>> {
        self.import(source).wait()
    }
}
```

**Comment by alteous** [comment]
So, I've been experimenting with the 'everything is async' idea (see above) and it works out quite nicely. Surprisingly, the code is more pleasant to write than its synchronous version.

Making everything async has a downside though, namely it's difficult to implement zero-copy deserialisation. I'm keen to provide this feature since it's something that only Rust can perform feasibly due to its memory safety guarantees.

In order to provide `ZeroCopyImporter` the relevant import code may have to be duplicated from `StaticImporter` with some tweaks. Since loading glTF is primarily an I/O bound task I'm tempted to leave zero-copy deserialisation out for the 1.0 release so this code duplication isn't necessary. Note that this could always be re-implemented at a later date if there's significant demand.

TL;DR: By making everything async I'm left with a choice:

 * Remove (temporarily) `ZeroCopyImporter`, or,
 * Maintain two import implementations.

**Comment by bwasty** [comment]
I like the idea of returning futures (though I haven't looked at the `AsyncSource` examples too closely). About async vs zero-copy: I think async has more potential for large speedups and 'one copy' doesn't sound too bad. So removing `ZeroCopyImporter` could be fine for the moment. And later when we have a full async example and some bigger, real scenes we could also benchmark too see if it's worth it.

Another potential aspect of async is multi-threaded uploading of data to GPU with one of the modern Graphics APIs (e.g. per `primitive`). Would this already be possible now?

And another thing about copying - at the moment I'm iterating all the geometry to fill `Vertex` structs that are then passed to OpenGL (without `multizip` though - with all attributes wrapped in `Option` I didn't see how to do it nicely). 
Ideally I'd directly pass a pointer to the buffer data, getting the size/offset/stride data from the `accessor`/`bufferView` - that is/will still be supported, right?
(At least that's how I'd do it in WebGL with `ArrayBuffer[View]s`. Come to think of it, with Web Workers a Javascript version could be be zero-copy AND async I think, just WebGL access is still limited to the 'main thread').

**Comment by alteous** [comment]
I totally agree with you about the async performance potential. Zero-copy could be benchmarked at a later date as suggested.

I think multi-threaded uploading is simple if all the buffers and images are read into memory ahead of time, but I'm still not convinced this is a good idea. If instead the buffers and images are lazily loaded then we have to ensure thread / memory safety of the loaded data. I think the following idea would work:

```rust
use std::sync::RwLock;

enum Lazy<T, E> {
    /// Data is available to be read.
    Ready(T),
    /// Data is being acquired on another thread.
    InProgress,
    /// Data is neither ready nor in progress.
    NotReady(BoxFuture<T, E>),
}

struct Gltf<'a> {
    buffers: Vec<RwLock<Lazy<BufferData, Error>>>,
    images: Vec<RwLock<Lazy<ImageData, Error>>>,
    // some fields omitted
}
```
The locks could impose a performance penalty, so a design like this would have to be considered carefully.

> Ideally I'd directly pass a pointer to the buffer data, getting the size/offset/stride data from the accessor/bufferView - that is/will still be supported, right?

Of course!

I think with more modern graphics APIs, such as Vulkan, the vertex layout needs to be known ahead of time, so iterating over the geometry is the only way. I'm not 100% sure of this though.

**Comment by alteous** [comment]
I'm starting to realise the crate may be trying to doing too much on behalf of the user. To give the user more freedom and to limit the scope of the crate, I've been considering a different approach:
```rust
/// Here to stay
trait Source {
    ...
}

/// Almost the same as the current `Gltf`, but the buffer and image data has shared ownership.
struct Gltf {
    /// No zero-copy anymore
    root: json::Root,

    /// Copy of user return type - quite complex to express.
    data: Data,
}

/// Same as `import::Error`.
enum Error<S: Source> { ... }

/// Replaces `Importer`. Is like a builder.
struct Import;

struct Preloaded<S: Source>(...);
impl<S: Source> Future for Preloaded<S> {
    type Item = (Gltf, Vec<Arc<Buffer>>, Vec<Arc<Image>>);
    type Error = Error<S>;
   ...
}

struct Lazy<S: Source>(...);
impl<S: Source> Future for Lazy<S> {
    type Item = (Gltf, Vec<Arc<Lazy<Buffer, S::Error>>>, Vec<Arc<Lazy<Image, S::Error>>>);
    type Error = Error<S>;
    ...
}

impl Import {
    /// Creates an unloaded glTF instance from the given custom source.

    /// Returns a `Future` that resolves to a glTF asset complete with its
    /// buffer and image data.
    pub fn preloaded<S: Source>(source: S, -> Preloaded<S> {
        ...
    }

    /// Returns a `Future` that resolves to a glTF asset with lazily-loaded
    /// buffer and image data.
    pub fn lazy<S: Source>(source: S) -> Lazy<S> {
        ...
    }
}
```
where `Lazy<T, E>` implements `Future<Item = T, Error = E>`.

This design has the advantage that buffer and image loading is left to the user whilst `Gltf` still has access to buffer and image data for generating normals, iterating over attributes, etc. The rather obvious downside is the crazy type stuff. Still, at least it's not as bad as [vulkano](https://twitter.com/tomaka17/status/832330520004530177).

**Comment by bwasty** [comment]
How well can the crazy types be hidden for the simple case where the user doesn't care about async/laziness/custom sources?

For me it would be fine I think. Or rather, I guess some craziness is hard to avoid for the advanced stuff I wanna try :)

---
> I think with more modern graphics APIs, such as Vulkan, the vertex layout needs to be known ahead of time, so iterating over the geometry is the only way. I'm not 100% sure of this though.

From what I read about `gfx-rs`, pipelines, which contain the vertex layout, can also be constructed at runtime, they just don't have a convenient API for it yet.

**Comment by alteous** [comment]
> How well can the crazy types be hidden for the simple case where the user doesn't care about async/laziness/custom sources?

Actually the types are not as bad as I originally thought. I found hidden in the docs this wonderful data structure called [`future::Shared`](https://docs.rs/futures/0.1.14/futures/future/trait.Future.html#method.shared) that hides all the ugliness of the lazy evaluation. Having everything async has made the crate much easier to design. The user can perform any combination of synchronous / asynchronous and eager / lazy loading without too much hassle.

> From what I read about gfx-rs, pipelines, which contain the vertex layout, can also be constructed at runtime, they just don't have a convenient API for it yet.

That's interesting, it's been that way for quite a while though. I remember when prototyping a glTF 1.0 renderer trying to do anything at runtime with `gfx-rs` was not at all convenient!

**Comment by alteous** [comment]
For visibility, I've pushed the [new (and hopefully final) strategy](https://github.com/alteous/gltf/blob/import2/src/gltf.rs#L36) to a temporary branch. Synopsis:

```rust
pub struct Gltf {
    buffers: Vec<Shared<BoxFuture<Box<[u8]>, import::Error>>>,
    images: Vec<Shared<BoxFuture<Box<[u8]>, import::Error>>>,
    root: root::Root,
}

pub trait Source: Send + 'static {
    fn source_gltf(&self) -> BoxFuture<Box<[u8]>, import::Error>;
    fn source_external_data(&self, uri: &str) -> BoxFuture<Box<[u8]>, import::Error>;
}
```

There is one rather annoying problem left to tackle, namely how user errors should be returned from the `Source` trait. I would really like to put a `Foo<'a, S: Source>` on everything but the resulting types are not convenient to use in parameter lists since, unlike the `'a` lifetimes, the type `S` can't be deduced by the compiler.

What I'd like to do:

```rust
pub trait Source: Send + 'static {
    type Error;
    fn source_gltf(&self) -> BoxFuture<Box<[u8]>, Self::Error>;
    fn source_external_data(&self, uri: &str) -> BoxFuture<Box<[u8]>, Self::Error>;
}
```

The problem...
```rust
pub struct Gltf<S: Source> {
    root: root::Root<S>,
    ...
}

pub struct Root<S: Source> { ... }

// etc.

fn process_mesh<'a>(mesh: gltf::Mesh<'a, gltf::import::source::from_path::FromPath>) {
    ...
}

// or

fn process_mesh<'a>(mesh: gltf::Mesh<'a, UserSourceImpl>) {
    ...
}
```

It's very inconvenient!

**Comment by alteous** [comment]
The `future::Shared` type should solve our lazy-loading needs if https://github.com/alexcrichton/futures-rs/pull/528 is merged.

**Comment by bwasty** [comment]
I ran into the confusing-as-hell case with incomprehensible trait-related compiler errors when dependency versions don't match - i.e. I added `alexcrichton/futures-rs` as a git dependency to my Cargo.toml (since your PR was merged already) and got this:
```
error[E0599]: no method named `wait_future` found for type `futures::task::Spawn<std::boxed::Box<futures::future::F
uture<Error=gltf::import::Error, Item=gltf::Gltf> + std::marker::Send>>` in the current scope
  --> src/main.rs:89:65
   |
89 |         let gltf = match spawn(gltf::import::from_path(source)).wait_future() {
   |                                                                 ^^^^^^^^^^^
   |
   = note: the method `wait_future` exists but the following trait bounds were not satisfied:
           `std::boxed::Box<futures::future::Future<Error=gltf::import::Error, Item=gltf::Gltf> + std::marker::Send
> : futures::Future`
```
Perhaps `gltf` could re-export `futures-rs` or a part of it? (for reference, similar case where I ran into this before: https://github.com/slack-rs/slack-rs-api/pull/42).

**Comment by bwasty** [comment]
Another case where the compiler was surprisingly unhelpful:
```
error[E0277]: the trait bound `futures::Future<Error=futures::future::SharedError<gltf::import::Error>, Item=gltf::
mesh::Positions> + std::marker::Send: std::marker::Sized` is not satisfied
  --> src/render/primitive.rs:78:58
   |
78 |         let positions = g_primitive.positions().unwrap().wait();
   |                                                          ^^^^ `futures::Future<Error=futures::future::SharedError<gltf::import::Error>, Item=gltf::mesh::Positions> + std::marker::Send` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `futures::Future<Error=futures::future::SharedError<gltf::import::Error>, Item=gltf::mesh::Positions> + std::marker::Send
```

Solved with 
```
use futures::Future;
```

I guess the only thing we could do here is a good example/a note in the docs.

**Comment by bwasty** [comment]
Finished a [quick and dirty update of my code](https://github.com/bwasty/gltf-viewer/commit/a9b33087bef7fd05e50849f873a5d97135ed8d9d) - now I'm hitting an assert at runtime: 
```
thread 'main' panicked at 'assertion failed: self.size() == mem::size_of::<T>()', /Users/bwasty/.cargo/git/checkout
s/gltf-745b33883171ad1d/f80db57/src/accessor.rs:73:8
```
Can be reproduced with `cargo run -- <path to any gltf>`

**Comment by alteous** [comment]
Thanks for reporting. `futures` is notorious for its compiler errors - more documentation and examples is probably the best remedy. Re-exporting `futures` is also worth consideration.

I really liked your idea (from a while ago) about loading data per-scene. There wouldn't be any need to call `wait()` for mesh data etc. in this case. The downside would be less choice in what gets loaded (although a client rendering glTF is likely to use all the data anyway) and hence an increase in memory consumption.

> now I'm hitting an assert at runtime

Oops, my bad! https://github.com/alteous/gltf/commit/b55250f36b9f8c7a14c79d489b46ac84413c2348 should fix this.

**Comment by bwasty** [comment]
> Oops, my bad! b55250f should fix this.

Getting there! Now something is rendering, but it's largely random garbage (and different on every start). Sometimes it resembles the model, sometimes not. Once the BoomBox was correct :)
Examples:
<img width="200" alt="screen shot 2017-07-07 at 18 24 30" src="https://user-images.githubusercontent.com/1647415/27967228-851fa8b0-6342-11e7-93b9-242dd8f1718b.png"><img width="200" alt="screen shot 2017-07-07 at 18 25 37" src="https://user-images.githubusercontent.com/1647415/27967229-85399f7c-6342-11e7-9092-02d88e6e5db3.png">
(had to comment out some more stuff that was not correctly adapted: https://github.com/bwasty/gltf-viewer/commit/7d4c0e636109df310b5e4870f2cea133ea9c11a4).

> I really liked your idea (from a while ago) about loading data per-scene. There wouldn't be any need to call wait() for mesh data etc. in this case.

It would be really nice not to `wait()` for every attribute/index access. My first thought here was to wait on the whole primitive and then have all its data available. Doing it on scene level would again limit parallelism a bit, wouldn't it?

**Comment by bwasty** [comment]
Now also tried release mode and the stable toolchain (I'm usually on nightly), but no difference. Another fun screeenshot: 
<img width="200" alt="screen shot 2017-07-07 at 18 51 07" src="https://user-images.githubusercontent.com/1647415/27968053-c67e2964-6345-11e7-9239-b570aed61f95.png">

**Comment by bwasty** [comment]
Perhaps also worth mentioning: Even before the async changes I had weird artifacts on files with a "large" number of nodes, such as `Buggy` or `VC`. Around 80% of each scene looked fine, the rest of the geometry was weirdly broken.

**Comment by bwasty** [comment]
> I've been keeping an eye on your viewer, which is coming along nicely. I'd recommend putting Cargo.lock in your .gitignore.

Why? It's good practice: http://doc.crates.io/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries

At work we don't have proper lock files for python and npm (yet) and that causes lots of pain occasionally (suddenly broken builds, can't deploy...)

**Comment by alteous** [comment]
> My first thought here was to wait on the whole primitive and then have all its data available. Doing it on scene level would again limit parallelism a bit, wouldn't it?

Yes, it probably would. I'm worried about the memory footprint too. Let's try per-primitive loading and see how it improves ergonomics.

I'm sorry about the 'weird artifacts'. They are most likely due to the crate incorrectly managing buffer data. It's difficult to unit test this so instead I'll clone your viewer and poke around to find the source of the artifacts.

> > I'd recommend putting Cargo.lock in your .gitignore.

> Why? It's good practice

I stand corrected. The recommendation is for library crates, not for binaries. Thanks for reminding me.

**Comment by alteous** [comment]
![lantern](https://user-images.githubusercontent.com/8549126/27974890-db5862fc-6357-11e7-952f-6747d3e5be0f.png)

It should be okay now. The issue was my inability to use standard library functions haha.

**Comment by bwasty** [comment]
Very nice! No need to be sorry, some bugs are expected at this stage :)

VC and Buggy are now again broken in the same way as before - no idea if the problem is in the viewer or the loader. I'll debug that later (first I want textures and a network-based `Source` - the trait is stable now, right?)

> It's difficult to unit test this ...

Always a tricky topic for 3D graphics :)
Not sure what can be done in the loader, but for the viewer I plan to add a screenshot mode (cli parameter) and an integration test that loads all sample models, screenshots and creates an HTML page. Perhaps I'll even add image diffing later.

**Comment by alteous** [comment]
> the trait is stable now, right?

It's "almost stable". I want to wait a little longer to make sure it's not going to have to change.

**Comment by alteous** [comment]
I'm nearly happy with the design of the `Source` trait but not with the way it's currently being used.

I think lazy-loading has added too much complexity to the import process, so preloading everything instead is tempting option. Let's try preloading everything and see how it scales with more data. We can always revisit this if necessary.

**Comment by bwasty** [comment]
If I understand this correctly - keep the parallel/async loading support, but don't do it lazy?

**Comment by alteous** [comment]
Yes, that's the idea.

**Comment by bwasty** [comment]
Sounds good. At this level, a separate function to only load a specific scene might make sense again (later).

**Comment by alteous** [comment]
I'm happy with the design of the `Source` trait and ready to close this issue. Final remarks are the `BoxFuture` return types may be changed to `impl Future` once that feature is stabilised - the same applies to the import code.

The proposed final design
```rust
/// A trait for representing sources of glTF data that may be read by an importer.
pub trait Source: Debug + Send + Sync + 'static {
    /// User error type.
    type Error: std::error::Error + Send + Sync;

    /// Read the contents of a .gltf or .glb file.
    fn source_gltf(&self) -> BoxFuture<Box<[u8]>, Self::Error>;

    /// Read the contents of external data.
    fn source_external_data(&self, uri: &str) -> BoxFuture<Box<[u8]>, Self::Error>;
}
```

And once `impl Trait` is stabilised...
```rust
/// A trait for representing sources of glTF data that may be read by an importer.
pub trait Source: Debug + Send + Sync + 'static {
    /// User error type.
    type Error: std::error::Error + Send + Sync;

    /// Read the contents of a .gltf or .glb file.
    fn source_gltf(&self) -> impl Future<Item = Box<[u8]>, Error = Self::Error>;

    /// Read the contents of external data.
    fn source_external_data(&self, uri: &str) -> impl Future<Item = Box<[u8]>, Error = Self::Error>;
}
```

If calling `source_gltf` or `source_external_data` fails then the error `import::Error::Source(...)` is returned.

**Comment by bwasty** [comment]
I've worked on implementing a "proper" multi-threaded HTTP Source and it works fine so far :)
With 1 thread the [VC model](https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Models/master/2.0/VC/glTF/VC.gltf) loads in 6.5 seconds, with 8 threads in 1.5 seconds (averaged).

One small problem that I don't quite understand is that my custom error type doesn't get shown fully - [annotated source](https://github.com/bwasty/gltf-viewer/blob/26faceb39517a76e99f5f735b55f50f06793c806/src/http_source.rs#L49-L53).

**Comment by alteous** [comment]
The HTTP source isn't async as [this call](https://github.com/bwasty/gltf-viewer/blob/26faceb39517a76e99f5f735b55f50f06793c806/src/http_source.rs#L56) is blocking. I would predict an async implementation on a single thread would have similar performance to the multi-threaded implementation. 

I'm not sure about the custom error type not displaying. I ran `cargo run -- https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Models/master/2.0/BrainStem/glTF/BrainStem` (missing .gltf extension) which output `Error: Source(HttpError("404 Not Found"))`.

**Comment by bwasty** [comment]
> The HTTP source isn't async as this call is blocking.

I know - I started with `reqwest` because it's simple and "just works" (`hyper` didn't when I tried it a while ago, before it was async). I though multi-threading is good enough, but now I'll have to try the new `hyper` and measure :)

> I'm not sure about the custom error type not displaying. I ran ...

Weird, the problem went away after I ran without incremental compilation once (normally I prefix all cargo commands with `CARGO_INCREMENTAL=1`).

**Comment by alteous** [comment]
The `Source` trait has a different purpose now. That is, to retrieve pre-loaded buffer data rather than I/O objects that load the buffer data. Voting to close.

**Comment by bwasty** [comment]
Before closing, I'd like to have a clear idea how to achieve the same things as with the previous `Source`. As far as I understand it would be:
* Load only the json
* Iterate the json, extracting all buffer/image uris
* Load them (async/parallel if desired) and put the results into a `Source` implementation
* Call `gltf::gltf::Gltf::loaded()`

Looking at the code, I'm confused about images. They're not handled at all as far as I see, so how can I get their data from a GLB file?

**Comment by alteous** [comment]
Nope. Try this instead:

 1. Load the JSON with `gltf_json::from_xxx()`.
 2. Create a `Gltf` wrapper with `Gltf::from_json()`.
 3. Iterate over the buffers with `Gltf::buffers()` and load as necessary.
 4. Put the loaded buffers into a `Source` implementation.
 5. Call `Gltf::loaded`.

Image data is no longer handled by the crate. It's entirely up to the user to decode. A `Loaded<Image>` however will return [`Image::Data`](https://docs.rs/gltf/0.8.5/gltf/image/enum.Data.html) which provides the buffer view or external buffer URI.

The [reference importer](https://github.com/alteous/gltf/blob/master/gltf-importer/src/lib.rs#L359) handles differences between .gltf and .glb files.

**Comment by bwasty** [comment]
But that's pretty much what I wrote :)
I deliberately included images since I don't want my I/O code to distinguish between buffers and images (it's all just file URLs in the end...).

The new structure makes the docs quite confusing. I looked at [gltf::image::Image](https://docs.rs/gltf/0.8.5/gltf/image/struct.Image.html), saw nothing about getting the actual data and concluded I'd have to do everything myself. I noticed my mistake when `Primitive` looked a bit too empty as well. I'd propose to add a very visible reference to `Loaded` on every wrapper struct doc string. Or will they all get a `loaded` method anyway?

I don't really like the `Image::Data` enum. Since it already deals with binary data in once case, I'd prefer to handle it as similar to buffers as possible - 'load' it and then get a `&[u8]` in either case (which has nothing to with decoding).

**Comment by alteous** [comment]
The docs are a bit confusing. It's not pleasant to browse every method of `Loaded` to find what you want. It would be great if we could get a reference to specific `Loaded` methods, but I'm not sure that's possible.

The two cases of `Image::Data` are subtly different. `FromBufferView` always has a MIME type, because otherwise it would be hard to know what encoding is being used. `External` has an optional MIME type, and when it doesn't exist the runtime must infer the encoding from the extension used.

**Comment by alteous** [comment]
In the next release, the semantics of the `Source` trait change from performing I/O to acquiring pre-existing data. It will also be moved to the upcoming and currently unstable `gltf-utils` crate.

Since `Source` will not required for the `gltf 1.0` release, I'm voting to close in favour of discussion in #78 instead.

**Comment by alteous** [comment]
Moving discussion to #78.
